crates/ruff_linter/src/rules/pycodestyle/rules/blank_lines.rs

```diff
@@ -0,0 +1,732 @@
+    }
+}
+
+/// ## What it does
+/// Checks for missing blank lines after end of function or class.
+///
+/// ## Why is this bad?
+/// PEP 8 recommends the using blank lines as following:
+/// - Two blank lines are expected between functions and classes
+/// - One blank line is expected between methods of a class.
+///
+/// ## Example
+/// ```python
+/// class User(object):
+///     pass
+/// user = User()
+/// ```
+///
+/// Use instead:
+/// ```python
+/// class User(object):
+///     pass
+///
+///
+/// user = User()
+/// ```
+///
+/// ## References
+/// - [PEP 8](https://peps.python.org/pep-0008/#blank-lines)
+/// - [Flake 8 rule](https://www.flake8rules.com/rules/E305.html)
+#[violation]
+pub struct BlankLinesAfterFunctionOrClass(pub u32);
+
+impl AlwaysFixableViolation for BlankLinesAfterFunctionOrClass {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let BlankLinesAfterFunctionOrClass(blank_lines) = self;
+        format!("expected 2 blank lines after class or function definition, found ({blank_lines})")
+    }
+
+    fn fix_title(&self) -> String {
+        "Add missing blank line(s)".to_string()
+    }
+}
+
+/// ## What it does
+/// Checks for for 1 blank line between nested functions/classes definitions.
+///
+/// ## Why is this bad?
+/// PEP 8 recommends the using blank lines as following:
+/// - Two blank lines are expected between functions and classes
+/// - One blank line is expected between methods of a class.
+///
+/// ## Example
+/// ```python
+/// def outer():
+///     def inner():
+///         pass
+///     def inner2():
+///         pass
+/// ```
+///
+/// Use instead:
+/// ```python
+/// def outer():
+///     def inner():
+///         pass
+///
+///     def inner2():
+///         pass
+/// ```
+///
+/// ## References
+/// - [PEP 8](https://peps.python.org/pep-0008/#blank-lines)
+/// - [Flake 8 rule](https://www.flake8rules.com/rules/E306.html)
+#[violation]
+pub struct BlankLinesBeforeNestedDefinition(pub u32);
+
+impl AlwaysFixableViolation for BlankLinesBeforeNestedDefinition {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let BlankLinesBeforeNestedDefinition(blank_lines) = self;
+        format!("Expected 1 blank line before a nested definition, found {blank_lines}")
+    }
+
+    fn fix_title(&self) -> String {
+        "Add missing blank line".to_string()
+    }
+}
+
+/// Returns `true` if the token is Async, Class or Def
+fn is_top_level_token(token: &Option<Tok>) -> bool {
+    matches!(&token, Some(Tok::Class | Tok::Def))
+}
+
+/// Returns `true` if the token is At, Async, Class or Def
+fn is_top_level_token_or_decorator(token: &Tok) -> bool {
+    matches!(&token, Tok::Class | Tok::Def | Tok::At)
+}
+
+#[derive(Debug)]
+struct LogicalLineInfo {
+    first_token: Tok,
+    first_token_range: TextRange,
+    last_token: Tok,
+    last_token_range: TextRange,
+    is_comment_only: bool,
+    is_docstring: bool,
+    indent_level: usize,
```

Comment:
Calling this field `indent_level` can be misleading because, from what I understand from the code, it stores the indentation length in columns. 

To explain the difference, let's take a look at the following example:

```python
def test(a):
    if a == 1:
        pass
```

The indentation of `pass` is 8 spaces but the indent-level is 2.

Reply:
I looked at the `indent_level` used when instantiating  the `LogicalLineInfo`, and it is 8 for `pass`. Where is it 2 ?

Reply:
The level is two because:

```python
def test(a):  # Level 0, no indentation 
    if a == 1:  # Level 1, the if statement is indented
        pass	 # Level 2, one level deeper than the if statement
```

So the level is only concern with how nested the statement is, not with the length of the indentation. 

It's probably a very poor analogy: Assume you use an elevator to get to level 3. The button is labeled 3 and does not go to 20 meters above ground level.